### ISSUE TITLE
modify_unit doesn't apply the "loyal" effect

### ISSUE THREAD

**Issue by jostephd** [wesnoth/wesnoth#4137]
`[modify_unit][modifications][effect]apply_to=loyal` doesn't set set the unit's upkeep to "loyal". Seen in EI S7a where Grug still has `upkeep="full"` after it joins Gweddry.

**Comment by gfgtdf** [comment]
> `[modify_unit][modifications][effect]`

This is wrong syntax it's is supposed to be `[modify_unit][object][effect]`

**Comment by jostephd** [comment]
I checked again and EI S7a does `[modify_unit][modifications][trait][effect]apply_to=loyal`. I omitted the [trait] in the first post. Sorry about that.

https://wiki.wesnoth.org/DirectActionsWML#.5Bmodify_unit.5D says "Accepts generally the syntax inside of wml unit variables created by [store_unit] which can be viewed in a savefile or by using the inspect command." The :inspect command in S9 shows `[modifications][object][effect]apply_to=loyal`. That's not exactly what you wrote but I assume it would work too.

zookeeper and I spoke about this yesterday. If I understood correctly he said the WML code in the scenario used to work. @ln-zookeeper, is that right?

**Comment by ln-zookeeper** [comment]
I don't know if that code actually used to work (IIRC it was added in 1.13). But yes, the wiki does specifically mention that [trait] should be added **without** [modifications], so if that solves the problem then the current WML is in fact mistaken; the documentation doesn't really explicitly say what the current WML should do, after all.

**Comment by gfgtdf** [comment]
> EI S7a does `[modify_unit][modifications][trait][effect]apply_to=loyal`

I'm not 100% sure what this does, it probably does a config-merge on the [modifications] tag ( meging the apply_to loyal with the first tarit that the uni has) but that's just a guess, also does to the config based implementation of modufy_unit it will probably not effect the units upkeep= field since Thai will be oweritten with the original upkeep= value when in unit in unstored.


In any case I recommend not to use ` [modify_unit][modifications]`

**Comment by CelticMinstrel** [comment]
Yeah, that'll do a config merge on the modifications tag, but that does mean it should actually work, as the loyal trait is now in the modifications and `[modify_unit]` unstores the unit at the end, which if I recall correctly should trigger a rebuild.

That said, it's probably better to omit the `[modifications]` tag in this case.

(Actually, I'm not quite sure what the default behaviour is, but is there a chance that the existing code overwrites the first effect tag in the first trait of the unit?)

**Comment by gfgtdf** [comment]
> Yeah, that'll do a config merge on the modifications tag, but that does mean it should actually work, as the loyal trait is now in the modifications and [modify_unit] unstores the unit at the end, which if I recall correctly should trigger a rebuild

Well.it first rebuilds and then overwrites everything with the data from the unstored unit, basically undoing the rebuild, except for data that is missing in the stored unit, in this case the 'stored unit' still has exlicitly `upkeep=full` because that's what the unit had when it was stored.

**Comment by stevecotton** [comment]
Is there any release of 1.14.x that this worked in?

This has been added to the 1.14.9 milestone, but it sounds like an API change that shouldn't be in a stable release, unless it's a regression that was introduced after 1.14.0.

**Comment by gfgtdf** [comment]
> Is there any release of 1.14.x that this worked in?

I dont think so

> This has been added to the 1.14.9 milestone, but it sounds like an API change that shouldn't be in a stable release

Yes i agree i would  actually close this with invalid, since it works as exepcted once one knows that [modify_unit] is basically a store+unstore_unit internally. although the wml code in EI S7a should of course be fixed.

**Comment by jostephd** [comment]
> since it works as exepcted once one knows that [modify_unit] is basically a store+unstore_unit internally.

The whole point of having APIs is so developers who use the API wouldn't have to worry about how it's implemented. An API should behave predictably to someone who has read the documentation in the wiki but _doesn't_ know what the implementation is. Please make that be the case here. I don't mind whether the syntax is made to work or declared invalid, so long as the wiki makes it clear. (And ideally the schema validator would catch uses of the invalid syntax, but I don't know if that's easy to do)

**Comment by CelticMinstrel** [comment]
I think the schema inherits `[modify_unit]` fron `[unit]`, so if that's the case it'd be difficult to catch uses of `[modifications]`. There could be a warning or error emitted by the Lua function if the tag is present, though.

As for the implementation of `[modify_unit]` in terms of store/unstore, I still believe that that should be changed.

**Comment by gfgtdf** [comment]
> As for the implementation of [modify_unit] in terms of store/unstore, I still believe that that should be changed.

I actually did change that for the most part (like 90% uf usecases), but some parts in particular those like  [modifications] that neother have a real usecase nor are easily to do in lua afaik still run the old code, what basicially happens is that there is a whitelist of supported tags/attributes and if all attributes/subtags are on that whitelist the new code is used otherwise the old code is used. Still the   new code has in the end the same effect as the old code even though it does not use store+unsotre anymore.

**Comment by stevecotton** [comment]
> I don't mind whether the syntax is made to work or declared invalid, so long as the wiki makes it clear.

https://wiki.wesnoth.org/index.php?title=DirectActionsWML&curid=1365&diff=63368&oldid=62225

**Comment by gfgtdf** [comment]
maybe it should mention that the correct way is to directly use [object], [effect], [trait] in [modify_unit].
(assuming this is what one wants)

**Comment by jostephd** [comment]
Thanks, all. I added an item to #4145 about fixing the WML in EI S7a. I guess we can close this issue now?

**Comment by CelticMinstrel** [comment]
`[effect]` in `[modify_unit]` isn't for permanent effects. It'll be reverted when the unit is rebuilt just like using keys directly in `[modify_unit]`. Basically it's like the old `[object]no_write=yes`.

**Comment by AI0867** [comment]
The case in HttT is that a trait is added to the unit, but the trait's effect (modifying the upkeep attribute) is not applied until the unit is rebuilt.

**Comment by CelticMinstrel** [comment]
So just to clarify… this doesn't work:

```
[modify_unit]
  [modifications]
    [trait]
      [effect]
        apply_to=loyal
      [/effect]
    [/trait]
  [/modifications]
[/modify_unit]
```

But these two do work:

```
[modify_unit]
  [trait]
    [effect]
      apply_to=loyal
    [/effect]
  [/trait]
[/modify_unit]

[modify_unit]
  [effect]
    apply_to=loyal
  [/effect]
[/modify_unit]
```

Is that accurate? If so, then I don't think we need to do anything.

**Comment by gfgtdf** [comment]
> Is that accurate?

Yes

> If so, then I don't think we need to do anything.

We need to update the wrong wml code if it's still in mainline.

**Comment by CelticMinstrel** [comment]
From earlier comments it looks like that was already addressed?

**Comment by stevecotton** [comment]
> We need to update the wrong wml code if it's still in mainline.

Reopening until we're checking it in the schema.

Known places where it still needs fixing:
* EI S7a's Grug (mentioned in #4145 )
* TSG S05's Urza (I'm currently fixing that as part of cherry-picking #5131 )

**Comment by CelticMinstrel** [comment]
In order to check it in the schema we probably need to alter the tag hierarchy… maybe making `[unit]` and `[modify_unit]` inherit from a common pseudo-tag?

**Comment by Wedge009** [comment]
Is this still applicable? After the redesign of EI, I'm not sure if the previous comments about Grug still applies for S7a. I've searched for instances of `apply_to=loyal` across the campaigns as well.
